sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java

```diff
@@ -53,20 +53,45 @@ public void skip() {
     throw new UnsupportedOperationException();
   }
 
+  private void updateCurrentByte() {
+    try {
+      currentByte = in.read();
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Failed to read a byte", e);
+    }
+  }
+
   @Override
   public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      c.putBoolean(rowId + i, readBoolean());
+    int i = 0;
+    if (bitOffset > 0) {
+      i = Math.min(8 - bitOffset, total);
+      c.putBooleans(rowId, i, currentByte, bitOffset);
+      bitOffset = (bitOffset + i) & 7;
+    }
+    for (; i + 7 < total; i += 8) {
+      updateCurrentByte();
+      c.putBooleans(rowId + i, currentByte);
+    }
+    if (i < total) {
+      updateCurrentByte();
+      bitOffset = total - i;
+      c.putBooleans(rowId + i, bitOffset, currentByte, 0);
     }
   }
 
   @Override
   public final void skipBooleans(int total) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      readBoolean();
+    int totalByte = total / 8;
+    if (totalByte > 0) {
+      try {
+        in.skipFully(totalByte - 1L);
```

Comment:
Why -1? If you skip 9 values, then you need to skip 1 byte and load another one to offset, isn't it?

Also, what happens when you read 7 values and then need to skip 11? How do you skip the last value in a current byte? Do you need to adjust for the current bit offset before calculating totalBytes and calling skipFully?

Reply:
Right, the first line should have been
```
 int totalByte = (bitOffset + total) / 8; 
```

`Why -1?`: Because we read another byte in the very next line and update `currentByte`.
E.g. skipping 9 values means `bitOffset = 0` and `total = 9`, then
```
 int totalByte = (bitOffset + total) / 8; // = (0+9)/8 = 1
...
        in.skipFully(totalBytes - 1L); // in.skipFully(0)
        currentByte = in.read(); // making currentByte ready
```

`read 7 values and ... skip 11?`: For this example, `bitOffset = 7` is given and running `skipBooleans(total = 11)` will be
```
 int totalByte = (bitOffset + total) / 8; // = (7+11)/8 = 2
...
        in.skipFully(totalBytes - 1L); // in.skipFully(1)
        currentByte = in.read(); // making currentByte ready
...
bitOffset = (bitOffset + total) & 7; // = (7+11)&7 = 2
```

Reply:
Updated the logic.